File: autotest/src/tests/src-rs/src/test_map.rs
// SPDX-FileCopyrightText: 2025 ANSSI
// SPDX-License-Identifier: Apache-2.0

use crate::device_utils::get_device_by_name;
use crate::devices::*;
use crate::test_log::*;
use sentry_uapi::status::Status;
use sentry_uapi::systypes::*;
use sentry_uapi::*;

pub fn test_map() -> bool {
    test_suite_start!("sys_map");
    let mut ok = true;
    ok &= test_map_mapunmap();
    ok &= test_map_invalidmap();
    ok &= test_map_unmap_notmapped();
    test_suite_end!("sys_map");
    ok
}

fn test_map_unmap_notmapped() -> bool {
    test_start!();
    // This will fail if the i2c1 is not found
    let device = get_device_by_name("i2c1").expect("i2c1 device not found");
    let mut dev: DeviceHandle = 0;
    let ok = check_eq!(__sys_get_device_handle(device.id as u8), Status::Ok)
        & unsafe {
            copy_from_kernel(
                &mut dev as *mut _ as *mut u8,
                core::mem::size_of::<DeviceHandle>(),
            )
        }
        == Status::Ok & check_eq!(__sys_unmap_dev(dev), Status::Invalid);
    test_end!();
    ok
}

fn test_map_invalidmap() -> bool {
    test_start!();
    let device = get_device_by_name("i2c1").expect("i2c1 device not found");
    let mut dev: DeviceHandle = 0;
    let ok = check_eq!(__sys_get_device_handle(device.id as u8), Status::Ok)
        & unsafe {
            copy_from_kernel(
                &mut dev as *mut _ as *mut u8,
                core::mem::size_of::<DeviceHandle>(),
            )
        }
        == Status::Ok;
    let invalid_dev = dev.wrapping_add(42);
    let ok = ok & check_eq!(__sys_map_dev(invalid_dev), Status::Invalid);
    test_end!();
    ok
}

fn test_map_mapunmap() -> bool {
    test_start!();
    let device = get_device_by_name("i2c1").expect("i2c1 device not found");
    let mut dev: DeviceHandle = 0;
    let mut ok = check_eq!(__sys_get_device_handle(device.id as u8), Status::Ok)
        & unsafe {
            copy_from_kernel(
                &mut dev as *mut _ as *mut u8,
                core::mem::size_of::<DeviceHandle>(),
            )
        }
        == Status::Ok;
    log_info!("handle is {:#x}", dev);
    ok &= check_eq!(__sys_map_dev(dev), Status::Ok);

    #[cfg(CONFIG_ARCH_MCU_STM32U5A5)]
    if ok {
        log_info!("device mapped, checking registers");
        let base = device.baseaddr;
        for offset in (0..12 * 4).step_by(4) {
            let val = unsafe { core::ptr::read_volatile((base + offset as usize) as *const u32) };
            if offset != 6 * 4 {
                ok &= check_eq!(val, 0x0);
            } else {
                ok &= check_eq!(val, 0x1);
            }
        }
    }

    log_info!("unmapping");
    ok &= check_eq!(__sys_unmap_dev(dev), Status::Ok);
    test_end!();
    ok
}

File: autotest/src/tests/src/test_map.c
// SPDX-FileCopyrightText: 2023 Ledger SAS
// SPDX-License-Identifier: Apache-2.0

#include <inttypes.h>
#include <test_log.h>
#include <test_assert.h>
#include <uapi/uapi.h>
#include <devices-dt.h>
#include "test_gpio.h"

#if DEVICE_LIST_SIZE > 0

void test_map_unmap_notmapped(void) {
    Status res;
    devh_t dev;
    TEST_START();
    res = __sys_get_device_handle((uint8_t)devices[DEV_ID_I2C1].id);
    copy_from_kernel((uint8_t*)&dev, sizeof(devh_t));
    ASSERT_EQ(res, STATUS_OK);
    res = __sys_unmap_dev(dev);
    ASSERT_EQ(res, STATUS_INVALID);
    TEST_END();
}

void test_map_invalidmap(void) {
    Status res;
    devh_t dev;
    TEST_START();
    res = __sys_get_device_handle((uint8_t)devices[DEV_ID_I2C1].id);
    copy_from_kernel((uint8_t*)&dev, sizeof(devh_t));
    ASSERT_EQ(res, STATUS_OK);
    dev += 42;
    res = __sys_map_dev(dev);
    ASSERT_EQ(res, STATUS_INVALID);
    TEST_END();
}

void test_map_mapunmap(void) {
    Status res;
    devh_t dev;

    TEST_START();
    res = __sys_get_device_handle((uint8_t)devices[DEV_ID_I2C1].id);
    copy_from_kernel((uint8_t*)&dev, sizeof(devh_t));
    ASSERT_EQ(res, STATUS_OK);
    LOG("handle is %lx", dev);
    res = __sys_map_dev(dev);
    ASSERT_EQ(res, STATUS_OK);
    if (res == STATUS_OK) {
        LOG("device mapped, checking registers");
        for (uint8_t offset = 0; offset < 12*sizeof(uint32_t); offset += 4) {
#ifdef CONFIG_ARCH_MCU_STM32U5A5
            /* checking effective registers reset values. Must match the datasheet */
            if (offset != 6*sizeof(uint32_t)) {
                ASSERT_EQ((uint32_t)*(uint32_t*)(devices[DEV_ID_I2C1].baseaddr + offset), 0x0);
            } else {
                ASSERT_EQ((uint32_t)*(uint32_t*)(devices[DEV_ID_I2C1].baseaddr + offset), 0x1);
            }
#endif
        }
    }
    LOG("unmapping");
    res = __sys_unmap_dev(dev);
    ASSERT_EQ(res, STATUS_OK);
    TEST_END();
}
#endif


void test_map(void) {
#if DEVICE_LIST_SIZE > 0
    TEST_SUITE_START("sys_map");
    test_map_mapunmap();
    test_map_invalidmap();
    test_map_unmap_notmapped();
    TEST_SUITE_END("sys_map");
#endif
}
